### PR TITLE
Create tests for `GetNavigationSource` extension method

### PR DESCRIPTION
**Pull request by habbes**
Related to issue #1118 

Related to PR #1158 

I want to refactor the `ODataPathExtensions.GetNavigationSource` extension method to improve its performance, but that method doesn't have tests. This PR adds tests covering all supported scenarios. This will make it safer to refactor without breaking existing behaviour.

### CODE REVIEW

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs [hunk 1]
```diff
@@ -63,6 +58,88 @@ public void GetNavigationSource_ThrowsArgumentNull_Path()
...
+        [InlineData("Customers/NS.GetBestOrders()", "Orders", EdmNavigationSourceKind.EntitySet)]
+        [InlineData("VipCustomer", "VipCustomer", EdmNavigationSourceKind.Singleton)]
+        [InlineData("VipCustomer/Name", "VipCustomer", EdmNavigationSourceKind.Singleton)]
+        [InlineData("VipCustomer/Orders", "Orders", EdmNavigationSourceKind.EntitySet)]
+        [InlineData("VipCustomer/NS.IsUpgraded", null, EdmNavigationSourceKind.None)]
+        [InlineData("Customers(1)/NS.SpecialCustomer", "Customers", EdmNavigationSourceKind.EntitySet)]
+        [InlineData("VipCustomer/NS.SpecialCustomer", "VipCustomer", EdmNavigationSourceKind.Singleton)]
+        [InlineData("Customers/$count", null, EdmNavigationSourceKind.None)]
+        [InlineData("Customers(1)/Account/Amount/$value", null, EdmNavigationSourceKind.None)]
+        [InlineData("$batch", null, EdmNavigationSourceKind.None)]
```
Comment (wachugamaina) [inline test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:92]: Is there a way to label these tests so as as to show which test case is being covered?

Comment (habbes) [reply test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:92]: We mostly use comments for such cases. Is that fine?

**Review by wachugamaina — APPROVED** [review]

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs (hunk 1 above)
Comment (wachugamaina) [reply test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:92]: I suppose that works just as well.

Comment (gathogojr) [reply test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:92]: @wachugamaina @habbes FYI https://github.com/xunit/xunit/discussions/2532

@wachugamaina Curious, do you use NUnit in Graph Onboarding repos? Does NUnit support parameterized tests with labels?

Comment (habbes) [reply test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:92]: @gathogojr thanks for the tip. That `DisplayName` features in that thread will be part of xunit v3 which is not out yet.

**Review by wachugamaina — APPROVED** [review]

File: test/Microsoft.AspNetCore.OData.Tests/Models/CustomersModelWithInheritance.cs [hunk 2]
```diff
@@ -290,6 +290,31 @@ public CustomersModelWithInheritance()
             getOrder.AddParameter("orderId", intType);
             model.AddElement(getOrder);
 
+            // functions with entity set
```
Comment (ElizabethOkerio) [inline test/Microsoft.AspNetCore.OData.Tests/Models/CustomersModelWithInheritance.cs:293]: what does this mean? bound?

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs [hunk 3]
```diff
@@ -63,6 +58,111 @@ public void GetNavigationSource_ThrowsArgumentNull_Path()
...
+        [InlineData("Customers(1)/Name", "Customers", EdmNavigationSourceKind.EntitySet)]
+        // navigation property segment
+        [InlineData("Customers(1)/Orders", "Orders", EdmNavigationSourceKind.EntitySet)]
+        // key segment of a navigation property
+        [InlineData("Customers(1)/Orders(1)", "Orders", EdmNavigationSourceKind.EntitySet)]
+        // navigation property link segment
+        [InlineData("Customers(1)/Orders(1)/$ref", "Orders", EdmNavigationSourceKind.EntitySet)]
+        // property segment of a navigation property
+        [InlineData("Customers(1)/Orders(1)/Amount", "Orders", EdmNavigationSourceKind.EntitySet)]
+        // non-contained navigation property without an entity set should return UnknownEntitySet
```
Comment (ElizabethOkerio) [inline test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:76]:
```suggestion
        // contained navigation property without an entity set should return UnknownEntitySet
```

File: test/Microsoft.AspNetCore.OData.Tests/Models/CustomersModelWithInheritance.cs [hunk 4]
```diff
@@ -305,6 +330,31 @@ public CustomersModelWithInheritance()
...
             model.AddElement(isSpecialAllUpgraded);
 
+            // unbound function and imports
+            // function with entity set
+            EdmFunction getTopCustomers = new EdmFunction(
+                "NS",
+                "GetTopCustomers",
+                new EdmCollectionTypeReference(new EdmCollectionType(new EdmEntityTypeReference(customer, false))),
+                isBound: false,
+                entitySetPathExpression: customers.Path,
```
[the remark this answers is not in the document]
Comment (gathogojr) [reply test/Microsoft.AspNetCore.OData.Tests/Models/CustomersModelWithInheritance.cs:340]: Good question!

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs [hunk 5]
```diff
@@ -63,6 +58,111 @@ public void GetNavigationSource_ThrowsArgumentNull_Path()
...
+        [InlineData("MyOrders(1)/OrderLines(1)", "OrderLines", EdmNavigationSourceKind.ContainedEntitySet)]
+        [InlineData("MyOrders(1)/OrderLines(1)/$ref", "OrderLines", EdmNavigationSourceKind.ContainedEntitySet)]
+        // property segment with a complex type, returns navigation source of previous segment
+        [InlineData("Customers(1)/Account", "Customers", EdmNavigationSourceKind.EntitySet)]
+        // property segment of a complex type, returns navigation source of previous segment
+        [InlineData("Customers(1)/Account/Bank", "Customers", EdmNavigationSourceKind.EntitySet)]
+        // dynamic path segment should return null
+        [InlineData("Customers(1)/Account/DynamicProperty", null, EdmNavigationSourceKind.None)]
+        // unbound operation import with an entity set path should return the entity set
+        [InlineData("GetTopCustomers()", "Customers", EdmNavigationSourceKind.EntitySet)]
```
Comment (gathogojr) [inline test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:91]: Same question @ElizabethOkerio asked. How does an unbound operation import have a navigation source?

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs [hunk 6]
```diff
@@ -63,6 +58,111 @@ public void GetNavigationSource_ThrowsArgumentNull_Path()
...
+        // unbound operation import without an entity set path should return null
+        [InlineData("GetTotalSalesAmount()", null, EdmNavigationSourceKind.None)]
+        // bound operation without an entity set path should return null
+        [InlineData("Customers(1)/NS.IsUpgraded()", null, EdmNavigationSourceKind.None)]
+        [InlineData("Customers/NS.UpgradeAll()", null, EdmNavigationSourceKind.None)]
+        // bound operaiton with an entity set path should return the entity set
```
Comment (gathogojr) [inline test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:97]:
```suggestion
        // bound operation with an entity set path should return the entity set
```

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs [hunk 7]
```diff
@@ -63,6 +58,111 @@ public void GetNavigationSource_ThrowsArgumentNull_Path()
...
+            else
+            {
+                Assert.NotNull(navigationSource);
+                Assert.Equal(expectedNavigationSource, navigationSource.Name);
+                Assert.Equal(navigationSourceKind, navigationSource.NavigationSourceKind());
+            }
+        }
+
+        [Fact]
+        public void GetNavigationSource_WhenPathTemplateSegment_ReturnsNull()
```
Comment (gathogojr) [inline test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:139]:
```suggestion
        public void GetNavigationSource_ReturnsNull_ForPathTemplateSegment()
```
or
```csharp
public void GetNavigationSourceReturnsNullForPathTemplateSegment()
```

**Review by gathogojr — DISMISSED** [review]
Left a few comments; otherwise LGTM

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs [hunk 8]
```diff
@@ -63,6 +58,111 @@ public void GetNavigationSource_ThrowsArgumentNull_Path()
...
+        // non-contained navigation property without an entity set should return UnknownEntitySet
+        [InlineData("MyOrders(1)/NonContainedOrderLines", "NonContainedOrderLines", EdmNavigationSourceKind.UnknownEntitySet)]
+        [InlineData("MyOrders(1)/NonContainedOrderLines(1)", "NonContainedOrderLines", EdmNavigationSourceKind.UnknownEntitySet)]
+        [InlineData("MyOrders(1)/NonContainedOrderLines(1)/$ref", "NonContainedOrderLines", EdmNavigationSourceKind.UnknownEntitySet)]
+        // contained navigation property should return ContainedEntitySet
+        [InlineData("MyOrders(1)/OrderLines", "OrderLines", EdmNavigationSourceKind.ContainedEntitySet)]
+        [InlineData("MyOrders(1)/OrderLines(1)", "OrderLines", EdmNavigationSourceKind.ContainedEntitySet)]
+        [InlineData("MyOrders(1)/OrderLines(1)/$ref", "OrderLines", EdmNavigationSourceKind.ContainedEntitySet)]
+        // property segment with a complex type, returns navigation source of previous segment
+        [InlineData("Customers(1)/Account", "Customers", EdmNavigationSourceKind.EntitySet)]
```
Comment (ElizabethOkerio) [inline test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:85]: Is there a property segment here?

**Review by ElizabethOkerio — APPROVED** [review]

File: test/Microsoft.AspNetCore.OData.Tests/Models/CustomersModelWithInheritance.cs (hunk 2 above)
Comment (habbes) [reply test/Microsoft.AspNetCore.OData.Tests/Models/CustomersModelWithInheritance.cs:293]: Functions that have an entity set path set. I think the entity set path expression is linked to the collection that contains the return type for functions that return an entity.

File: test/Microsoft.AspNetCore.OData.Tests/Models/CustomersModelWithInheritance.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (habbes) [reply test/Microsoft.AspNetCore.OData.Tests/Models/CustomersModelWithInheritance.cs:340]: Yes. I believe it applies for functions that return an entity or entity collection.

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs (hunk 5 above)
Comment (habbes) [reply test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:91]: This means that the function returns entities from an entity set. At least that's what I gathered from reviewing existing implementation.

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs (hunk 8 above)
Comment (habbes) [reply test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:85]: `/Account` is the property segment. `Account` is a complex property. And so the navigation source here is the navigation source of the previous segment, which is the key segment `(1)`, whose navigation source is the `Customers` entity set.

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs (hunk 5 above)
Comment (gathogojr) [reply test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:91]: @habbes How is such a function configured?

File: test/Microsoft.AspNetCore.OData.Tests/Models/CustomersModelWithInheritance.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (habbes) [reply test/Microsoft.AspNetCore.OData.Tests/Models/CustomersModelWithInheritance.cs:340]: I've updated the comment to reflect that this refers to the entity set path.

**Review by ElizabethOkerio — APPROVED** [review]

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs (hunk 5 above)
Comment (gathogojr) [reply test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:91]: @habbes Got it!
```csharp
modelBuilder.Function("GetTopCustomers").ReturnsCollectionFromEntitySet<Customer>("Customers");
```

**Review by gathogojr — APPROVED** [review]

**Pull request merged**